updated quiet function in the displayhook.py

thethomask opened the pull request:
Updated quiet function in the displayhook.py to properly recognize comment and semicolon by using the tokenize module. Thanks to @takluyver for heavy input and assistance in writing this code.

Commit:
add test file for display hooks

koenvb commented:
@thethomask I guess something went wrong with this commit. As it is an empty file?

thethomask commented:
Wat. Empty? I don't know what's going on. I'll look into it.

takluyver commented:
@thethomask , `test_displayhook.py` is the file you already added in #8321. Do you want to add to that pull request the changes to `displayhook.py` that you put in [this gist](https://gist.github.com/thethomask/fb3a9a5a6dfb385d18ac).

Commit:
Update displayhook.py

Updated quiet function in the displayhook.py to properly recognize comment and semicolon by using the tokenize module.

thethomask commented:
@takluyver Yes. Should be there now.

takluyver commented on IPython/core/tests/test_displayhook.py:
```diff
@@ -0,0 +1 @@
+code
```
We'll need to merge manually to remove this file. (making a note so I remember)

takluyver commented on IPython/core/displayhook.py:
```diff
@@ -83,12 +84,18 @@ def check_for_underscore(self):
-            # some uses of ipshellembed may fail here
-            return False
+
+	cell = self.shell.history_manager.input_hist_parsed[self.prompt_count]
```
The code that was here before catches `IndexError`, in case `input_hist_parsed` doesn't have that many entries. I'm not sure if that can actually still happen, but we should probably still catch that error, just to be on the safe side.

Commit:
Update displayhook.py

Updated to catch IndexError

thethomask commented:
@takluyver updated

takluyver commented:
Thanks! The try/except actually needs to be around the `...input_hist_parsed[self.prompt_count]` line, though, not the bit that iterates through tokens.

Carreau commented on IPython/core/displayhook.py:
```diff
@@ -83,9 +84,19 @@ def check_for_underscore(self):
         # do not print output if input ends in ';'
+ 
+	cell = self.shell.history_manager.input_hist_parsed[self.prompt_count]
+        sio = io.StringIO(cell)
```
Small indent issue here `cell =` seem to miss one leading space.

Carreau commented:
and the travis error is `TabError: inconsistent use of tabs and spaces in indentation` so you might have a mixing of tabs and space I guess :-)

takluyver commented:
Oh, I hadn't spotted that. Yes, it should indented using four spaces per indent level.

damianavila commented on IPython/core/displayhook.py:
```diff
@@ -83,9 +84,19 @@ def check_for_underscore(self):
+    	    for token in reversed(tokens):
+                if token.type in (tokenize.ENDMARKER, tokenize.COMMENT):
+            	    continue
+                if (token.type == tokenize.OP) and (token.string == ';'):
```
I would remove all the parenthesis on this line...

takluyver replied:
I tend to prefer the parentheses: I can never remember exactly which
operators have precedence, so explicit is better than implicit.
On 25 May 2015 5:21 pm, "Damián Avila" <notifications@github.com> wrote:

> In IPython/core/displayhook.py
> <https://github.com/ipython/ipython/pull/8457#discussion_r30998471>:
>
> >          try:
> > -            cell = self.shell.history_manager.input_hist_parsed[self.prompt_count]
> > -            return cell.rstrip().endswith(';')
> > +    	    for token in reversed(tokens):
> > +                if token.type in (tokenize.ENDMARKER, tokenize.COMMENT):
> > +            	    continue
> > +                if (token.type == tokenize.OP) and (token.string == ';'):
>
> I would remove all the parenthesis on this line...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipython/pull/8457/files#r30998471>.
>

damianavila replied:
not a big deal, if you prefer them, I am ok with that...

Carreau commented on IPython/core/displayhook.py:
```diff
@@ -83,9 +84,19 @@ def check_for_underscore(self):
         try:
-            cell = self.shell.history_manager.input_hist_parsed[self.prompt_count]
-            return cell.rstrip().endswith(';')
+    	    for token in reversed(tokens):
```
I think there is a tab/space inconsistency here too.
